[region-isolation] Add support for transferring parameters.

gottesmm opened the pull request:
A transferring parameter is a consuming like param modifier that can be used to transfer values to an async function such that you can transfer the value further.

Since it is consuming, one can assign into it. When one assigns into it, one has transferred the new value into the transferred parameter.

gottesmm committed:
[region-isolation] Add parsing/serialization/type system support for a transferring OwnershipSpecifier.

gottesmm committed:
[region-isolation] Implement checking for transferring parameters.

NOTE: This does not handle yet assignment into transferring parameters. In the
next commit, I am going to teach the checker that assigning into such a
parameter is a transfer.

gottesmm committed:
[region-isolation] Teach region isolation that assigning into a transferring parameter is a transfer of the value.

The specific semantics is if we assign into a transferring parameter's field,
then we "merge" src's value into the transferring parameter, so we
conservatively leave the region of the transferring parameter alone. If we
assign over the entire transferring parameter, we perform an assign fresh since
any value that used to be in the transferring parameter cannot reference
anything in its new value since they are all gone.

gottesmm commented:
@swift-ci smoke test

ktoso commented on test/Concurrency/transfernonsendable_strong_transferring.swift:
```diff
@@ -0,0 +1,228 @@
+
+func testTransferringParameter_canTransfer(_ x: transferring Klass, _ y: Klass) async {
+  await transferToMain(x)
+  await transferToMain(y) // expected-warning {{call site passes `self` or a non-sendable argument of this function to another thread, potentially yielding a race with the caller}}
```
I'm wondering if the warning should be phrased in terms of actor boundary crossing and not "to another thread", or is this intentional?

gottesmm replied:
I agree that said diagnostic needs to be wordsmithed (and actually split up into 1-2 other ones), this PR is just bringing up transferring, not solving that problem.

ktoso replied:
Yea not a blocker for merging ofc, just something I noticed

gottesmm replied:
Thanks for bringing it up = ).

gottesmm commented:
apple/swift-syntax#2410

@swift-ci smoke test

gottesmm commented:
https://github.com/apple/swift-syntax/pull/2410

@swift-ci smoke test

gottesmm commented:
(just rebasing/retesting since I am back from vacation)

gottesmm commented:
Looks like the windows/macOS failures are related to swiftpm

gottesmm committed:
[region-isolation] Eliminate some UB caused by dereferencing unchecked optionals.

llvm::Optional<T> used to make it so that in asserts builds if one dereferenced the optional and nothing was there, one would get an assert. std::optional<T> does not have that property.

gottesmm committed:
[region-isolation] Teach the checker that a use of a local value after being strongly transferred is an error.

Before the previous patch, we were just getting lucky on macOS due to UB. Now
that the UB is fixed, we correctly crash without this commit since we were not
pattern matching the simple case of a local value that was transferred and used
later.

gottesmm commented:
I got feedback from some people about some small changes about loosening the rules around assigning into a transferring parameter. I am going to do that in a follow on commit. This at least gets in the basics and ensures that I do not hit more merge conflicts around Serialization.
